src/modules/formatter.rs

```diff
@@ -0,0 +1,73 @@
+use std::{io::{BufWriter, Write}, process::{Command, Stdio}};
+
+
+/// This mechanism is built to support multiple formatters.
+/// 
+/// The idea is that amber should find the one installed, verify that its compatible and use the best one possible.
+#[derive(Debug, Clone, Copy)]
+#[allow(non_camel_case_types)]
+pub enum BashFormatter {
+    /// https://github.com/mvdan/sh
+    shfmt
+}
+
+impl BashFormatter {
+    /// Get all available formatters, ordered by: best ones at the start, worser at the end
+    pub fn get_all() -> Vec<BashFormatter> {
+        vec![
+            BashFormatter::shfmt
+        ]
+    }
+
+    /// Get available formatter
+    pub fn get_available() -> Option<BashFormatter> {
+        let all = Self::get_all();
+        all.iter().find(|x| x.is_available()).map(|x| *x)
+    }
+
+    /// Check if current formatter is present in $PATH
+    pub fn is_available(self: &Self) -> bool {
+        match self {
+            BashFormatter::shfmt =>
+                Command::new("shfmt")
+                    .arg("--version")
+                    .stdout(Stdio::null())
+                    .stderr(Stdio::null())
+                    .spawn()
+                    .map(|mut x| x.wait())
+                    .is_ok()
+        }
+    }
+
+    #[allow(dead_code)] // used in tests
+    pub fn as_cmd<T: From<&'static str>>(self: &Self) -> T {
+        match self {
+            BashFormatter::shfmt => "shfmt".into()
+        }
+    }
+
+    /// Format code using the formatter
+    pub fn format(self: &Self, code: String) -> String {
+        match self {
+            BashFormatter::shfmt => {
+                let mut command = Command::new("shfmt")
+                    .stdout(Stdio::piped())
+                    .stdin(Stdio::piped())
+                    .arg("-i").arg("4") // indentation
+                    .arg("-ln").arg("bash") // language
+                    .spawn().expect("Couldn't spawn shfmt");
```

Comment:
Are you sure that every version of shfmt supports given arguments?

I think that each formatter should be in their own struct with `format` trait function and we will just call it here. Logic specific to a given formatter will be handled there.

Reply:
We have to consider also that right now we implemented just because we don't have time to generate without the formatter a good bash so in the future the support will be removed.

Reply:
You don't know who is running amber on their comuter. It might be someone, who installed this formatter 5 years ago, and we're using some feature that was implemented 2 years ago. If I see correctly first stable release was in [December of 2016](https://github.com/mvdan/sh/releases?page=5).

We love to say `TODO: make this better` and we never do XD Given that we have a lot more important stuff to do I think that this could be here for a while. Let's not allow it to bite us later

Reply:
Maybe I'm too nagging here. I just think that it would look better and it's not a lot of work. But if others think it's good enough, I yield

Reply:
To solve this issue we could also have a link in documentation to the concrete version of `shfmt` that is supported in the github releases. We could show a warning that given version may not be compatible with our compiler